src/memo/memo.ml

```diff
@@ -704,16 +680,150 @@ module Call_stack = struct
+              incr Counters.edges_in_cycle_detection_graph;
+            let not_traversed_before =
+              Dag.Id.Set.is_empty children_added_to_dag
+            in
+            Stack_frame_with_state.record_child_added_to_dag frame ~dag_node_id;
+            match not_traversed_before with
+            | true -> add_path_impl stack caller_dag_node
+            | false ->
+              (* Same optimisation as above: no need to traverse again. *)
+              Ok ())))
+    in
+    add_path_impl stack (Lazy.force dag_node)
+
+  (* Add a dependency on the [dep_node] from the caller, if there is one. *)
+  let add_dep_from_caller =
+    let get_call_stack_tip () = get_call_stack () >>| List.hd_opt in
+    fun dep_node ->
+      let+ caller = get_call_stack_tip () in
+      match caller with
+      | None -> ()
+      | Some caller -> Stack_frame_with_state.add_dep caller ~dep_node
+end
+
+(* This module contains the essence of our cycle detection algorithm. Briefly,
+   the idea is as follows: whenever we are about to get blocked to wait for the
+   result of a computation that is currently running, we add the current call
+   stack to the cycle detection DAG. If this creates a cycle, we stop and report
+   a "dependency cycle" error; otherwise, we proceed with the blocking.
+
+   Below are some notes on how/why this algorithm works.
+
+   By "computation", we mean execution of a fiber. To share computation results
+   between multiple concurrent readers, we use [Once.t], which is a thin wrapper
+   around [Fiber.t] that executes the underlying fiber at most once. The first
+   reader "forces" the fiber and eventually gets the result, while other readers
+   subscribe to the future result of the computation via a [Fiber.Ivar.t]. This
+   blocks their execution and can lead to a deadlock if there is a dependency
+   cycle between different computations. To check for cycles *before* getting
+   blocked, a reader can use [Once.force_with_blocking_check].
+
+   One simple algorithm to check for cycles is to create a DAG node for every
+   computation, and add a DAG edge whenever a computation would like to read the
+   result of another unfinished computation. If adding an edge creates a cycle,
+   then we stop and report an error instead of getting blocked and deadlocked.
+   This algorithm is simple and it works, and Memo used it in the past. However,
+   the resulting DAG was often large, and so cycle detection was taking ~35% of
+   incremental zero rebuilds. As an optimisation, we developed another algorithm
+   described below.
+
+   The DAG produced by the above algorithm can contain many uninteresting nodes
+   and edges. For example, every node will have at least one incoming edge that
+   is added when the corresponding computation was initially "forced". But these
+   "forcing edges" cannot cause deadlocks by themselves; in fact, if there is no
+   parallelism, there may be no "blocking edges" at all, and the above algorithm
```

Comment:
There can be a cycle without any parallelism, and that does involve a blocking edge in our implementation, so this sentence seems a bit misleading. I'm not sure how to make this point precise, actually.

I think the purpose of this paragraph is to say that forcing edges are harmless. When we talked about before, I think we justified it like this:

```
But these "forcing edges" cannot cause deadlocks by themselves, since the fiber that forces a computation is not blocked, so it will keep making progress until it encounters a blocking edge.
```

The no-parallelism example is interesting, but I don't know if it does much to support this.

Reply:
Agreed, done.